### PR TITLE
JsonConverter- Sanitize Exception.Data dictionary keys option (on by default)

### LINKED ISSUE #2474: Json Serializer should sanitize custom dictionary keys
Usually NLog should just do the garbage-in and garbage-out, and not worry about broken data.

But when logging exceptions, then it is not always possible to control the Exception-data, as they can come from an external source.

The Exception.Data collection can contain a dictionary, where the dictionary-keys doesn't match normal json-properties (containing "." or "-"). This can be seen with Microsoft SQL exceptions.

### PR DESCRIPTION
**Pull request by snakefoot**
Attempt to resolve #2474

### CODE REVIEW

File: src/NLog/Targets/JsonSerializeOptions.cs [hunk 1]
```diff
@@ -69,6 +69,12 @@ public class JsonSerializeOptions
         [DefaultValue(false)]
         public bool EnumAsInteger { get; set; }
 
+        /// <summary>
+        /// Should dictionary keys be sanitized to match C-language variable name restriction (Replaces dots and hyphens and spaces with underscore)
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/JsonSerializeOptions.cs]: From [MSDN](https://msdn.microsoft.com/en-us/library/gg615485%28v=vs.88%29.aspx):

> The first character of a variable name must be either a letter, an underscore character (_), or the at symbol (@). Subsequent characters may be letters, underscore characters, or numbers.

[the remark this answers is not in the document]
Comment (304NotModified) [reply src/NLog/Targets/JsonSerializeOptions.cs]: please add that, thx :)

File: src/NLog/Targets/DefaultJsonSerializer.cs [hunk 2]
```diff
@@ -296,6 +286,31 @@ private void SerializeDictionaryObject(IDictionary value, StringBuilder destinat
                 }
                 else
                 {
+                    if (options.SanitizeDictionaryKeys)
+                    {
+                        int keyStartIndex = originalLength + (first ? 0 : 1) + (options.QuoteKeys ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/DefaultJsonSerializer.cs]: Moved into separate method.

File: src/NLog/Targets/DefaultJsonSerializer.cs [hunk 3]
```diff
@@ -296,6 +286,31 @@ private void SerializeDictionaryObject(IDictionary value, StringBuilder destinat
...
+                        if (destination.Length <= (keyStartIndex + (options.QuoteKeys ? 1 : 0)))
```
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/DefaultJsonSerializer.cs]: Done

File: src/NLog/Targets/JsonSerializeOptions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (snakefoot) [reply src/NLog/Targets/JsonSerializeOptions.cs]: Updated the summary-docs

**Pull request merged**